roboteam_ai/src/world/World.cpp

```diff
@@ -0,0 +1,305 @@
+    allRobots.insert(allRobots.end(), worldDataPtr->them.begin(), worldDataPtr->them.end());
+    return allRobots;
+}
+
+const std::vector<Robot> World::getUs() {
+    std::lock_guard<std::mutex> lock(worldMutex);
+    return worldDataPtr->us;
+}
+
+const std::vector<Robot> World::getThem() {
+    std::lock_guard<std::mutex> lock(worldMutex);
+    return worldDataPtr->them;
+}
+
+Robot World::getRobotClosestToPoint(const Vector2 &point, std::vector<Robot> robots) {
+
+    if (robots.empty())
+        return {};
+
+    unsigned int bestIndex = 0;
+    double closestDistance = 9999999;
+    double distanceToCheck;
+    for (unsigned int i = 0; i < robots.size(); i++) {
+        distanceToCheck = (robots[i].pos - point).length();
+        if (distanceToCheck < closestDistance) {
+            closestDistance = distanceToCheck;
+            bestIndex = i;
+        }
+    }
+
+    return robots[bestIndex];
+}
+
+Robot World::getRobotClosestToPoint(const Vector2 &point, WhichRobots whichRobots) {
+    std::vector<Robot> robotsCopy;
+    {
+        std::lock_guard<std::mutex> lock(worldMutex);
+
+        switch (whichRobots) {
+        case OUR_ROBOTS:
+            robotsCopy = worldDataPtr->us;
+            break;
+        case THEIR_ROBOTS:
+            robotsCopy = worldDataPtr->them;
+            break;
+        case ALL_ROBOTS:
+        default:
+            robotsCopy.insert(robotsCopy.end(), worldDataPtr->us.begin(), worldDataPtr->us.end());
+            robotsCopy.insert(robotsCopy.end(), worldDataPtr->them.begin(), worldDataPtr->them.end());
+            break;
+        }
+    }
+
+    getRobotClosestToPoint(point, robotsCopy);
+}
+
+Robot World::getRobotClosestToRobot(const RobotPtr &robot, WhichRobots whichRobots) {
+    return getRobotClosestToPoint(robot->pos, whichRobots);
+}
+
+Robot World::getRobotClosestToRobot(int id, bool ourTeam, WhichRobots whichRobots) {
+    RobotPtr robot = getRobotForId(id, ourTeam);
+    if (!robot)
+        return {};
+
+    return getRobotClosestToPoint(robot->pos, whichRobots);
+}
+
+Robot World::getRobotClosestToBall(WhichRobots whichRobots) {
+    Vector2 ballPos;
+    {
+        std::lock_guard<std::mutex> lock(worldMutex);
+        ballPos = worldDataPtr->ball.pos;
+    }
+
+    return getRobotClosestToPoint(ballPos, whichRobots);
+}
+
+bool World::robotHasBall(int id, bool ourTeam, double maxDist) {
+    return ourTeam ? ourRobotHasBall(id, maxDist) : theirRobotHasBall(id, maxDist);
+}
+
+bool World::ourRobotHasBall(int id, double maxDist) {
+    auto robot = getRobotForId(id, true);
+    if (!robot)
+        return false;
+
+    return robot->hasBall(maxDist);
+}
+
+bool World::theirRobotHasBall(int id, double maxDist) {
+    auto robot = getRobotForId(id, false);
+    if (!robot)
+        return false;
+
+    return robot->hasBall(maxDist);
+}
+
+const World::RobotPtr World::whichRobotHasBall() {
+    auto allRobots = getAllRobots();
+    double bestDistance = 999999;
+    Robot bestRobot = {};
```

Comment:
You should initialize the best robot's id as -1 if you want the check at the end to actually work. This as such will not work properly.

Reply:
nevermind, see robot constructor